README.md
# PyBluesky
### Version : 1.0.6


Author : Lakhya Jyoti Nath (ljnath)<br>
Date : April 2020 - May 2020<br>
Email : ljnath@ljnath.com<br>
Website : https://www.ljnath.com


[![Download PyBluesky](https://img.shields.io/sourceforge/dt/ljnath.svg)](https://sourceforge.net/projects/ljnath/files/latest/download)
[![GitHub license](https://img.shields.io/github/license/ljnath/PyBluesky)](https://github.com/ljnath/PyBluesky/blob/master/LICENSE)
[![GitHub stars](https://img.shields.io/github/stars/ljnath/PyBluesky)](https://github.com/ljnath/PyBluesky/stargazers)

[![Download PyBluesky](https://sourceforge.net/sflogo.php?type=13&group_id=3215162)](https://sourceforge.net/p/ljnath/)  

</br>
</br>

## INTRODUCTION
PyBluesky is a simple 2D python game developed using the pygame framework.</br>
Based on https://realpython.com/blog/python/pygame-a-primer
</br></br>

## DEMO
<img src="demo.gif" aligh="center"/>
<br><br>

## GAME MECHANICS
The game is simple where the objective is to navigate and shoot your way through the sky.
There are enemy missiles which travels from right-to-left with varied speed. These enemy missiles can be destroyed by shooting at them. With increase in game level, SAM launchers also moves on the ground, which can fire targeted missile at the jet. These missiles cannot be destroyed, so user needs to evade them.

The jet can be controlled both my keyboard and mouse. It can move in all directions viz. forward, backward, up and down. It can also be navigated diagonally with key combinations.
For example, in order to move in north-east directory, you can press both the RIGHT and UP arrow key.

While playing with keyboard, you can shoot bullets using the SPACEBAR and you can use your LEFT mouse button while playing with mouse.<br>

The gameplay has levels, which changes every 20 seconds. A level increase results in increases of enemy missiles.
It also gives 50 new ammo to the jet as well as the game score is bumped up by 10 points.

The game also features a power-up star which falls across the sky at each level.
Catching the power-up star will destroy all the enemy bullets in the current game frame.
</br></br>

## HOW TO PLAY
You can either download these code and directly play from it. For this you need to have python3 and the dependencies  mentioned in the requirements.txt file needs to be installed.

Alternately you can download and install the game binary from [sourceforge.com](https://sourceforge.net/p/ljnath/). This does not need any pre-requisite, as all the required pre-requisites are built into the binary package.
</br></br>

## LEADERBOARD
The game also features a network-controlled leaderboard. User scores along with few other metadata are published to a remote server.

During the game startup, the updated scores are download from the server and displayed as leaderboard.
</br></br>

## DEVELOPMENT

Following are the required dependencies for building the binary of this game.

- `pip install cx-Freeze==6.1` for creating distribution
- `sudo apt install zlib1g-dev`  for cx-Freeze installaion in ubuntu
</br></br>


CHANGELOG.md
# Changelog

## [1.0.6] - unreleased
- updated pygame library to v2.0


## [1.0.5] - 2021-06-10
- changed default quit choice to No in ExitMenu
- added support to hide the exit menu if ESCAPE key is pressed
- code refactoring

## [1.0.4] - 2020-05-17
- added exit prompt in the game, this also acts as a game pause feature
- changed size of sam missile
- changed the probabity of a sam-launcher firing from 50% to 33%
- changed the sam speed increase, now speed increases by 1 every 2 level instead of 1 level
- changes the jet speed based on input mode; in mouse mode jet speed it slightly faster
- changed the cloud and vegetation speed by increasing the vegetation's speed and reducing the cloud's speed
- replaces threading in network communication with asyncio
- renamed ReplayText to ReplayMenuText

## [1.0.3] - 2020-05-10
- added support for new api
- added sam launcher, they appear after level 5 and fire targeted missiles
- added support to add 10 pts on each level up
- added support to calulate player accuracy
- added support to upload offline scores when internet connectivity is available thus so scores are lost
- added common logging and serilization for both network-handler and leaderboard-handler
- added ground and grass to the vegetation
- added support to automatically open the update URL incase an update is available
- moved all gama data files under the /data/ directory
- changed log file name to game.log
- changed API key handling from network-handler to main game file
- changes game asset names to lowercase for UNIX comaptibity and changed name to be more meaningful
- changes game asset path name for UNIX comaptibity
- fixed hint text in player name input screen
- fixed invalid player name input via file
- fixed unwanted library imports in game classes
- removed unused game assets
- removed README file from final game distrubution build

## [1.0.2] - 2020-05-04
- added support for player name, games prompts to enter the name for the 1st time and saving to file
- added support for automatically loading exitsing player name from file
- added support for player name to be used for submitting game result
- added special hint in the input name menu
- added support to save game level data along with game score
- added support to boost enemy missle speed by 5% after each 10th level
- changed starting missile count from 4 to 2 
- changed missile increase factor per level  to half
- changed input mode selection message in gamemenu sprite
- changed jet speed from 8 to 7
- changes TitleScreen enum to Screen as it has all type of screen information
- changed bullet speed from 5 to 7, to avoid jet hitting its own bullet
- changed game icon, shortcut icon
- fixed leaderboard rendereing issue related to alignment
- fixed help message
- replace duplicate codes with list comprehension
- removed capturing of system username

## [1.0.1] - 2020-05-03
- added leaderboard functionality (requires internet)
- added game help menu
- added support to toggle between menu in game start and gameover screen
- added support to upload game score and current system user to remote api
- added support to check for game update at startup
- added support for clouds to notify user about new game update
- added support for dynamic menu spries creation instead for pre-created sprites
- changed main game music
- removed game sound for jet move up and move down (no more licenses)

## [1.0.0] - 2020-05-02
- renamed game to PyBluesky
- added background vegetation support
- added packaging support

## [0.12] - 2020-05-02
- updated sprite for bullet, jet
- added support for flashing powerup star, flash rate is configurable in the class
- added music which powerup is activated
- added sprite for deactivated missiles to distinguishing between activated ones

## [0.11] - 2020-05-01
- added powerup star which will mark all active missiles as bad
- added suport for collision detection between the jet and its bullet
- consuming powerup will increase game score by 100

## [0.10] - 2020-05-01
- added ammo limit, game starts with 100 ammo and 50 is added pn every level up (max ammo can be 999)
- added support for game to run in full screen by default
- added support to dynamically set game resolution as the monitor resolution
- channged score sprite color to dark grey
- change levelup time from 10s to 20s

## [0.9] - 2020-05-01
- added shoot functionality for the jet by clicking mouse of pressing spacebar
- updated score calculation, now destroying 1 missile gives 10 pts
- added game instruction in welcome screen
- added sound effects for shooting and destroying missiles

## [0.8] - 2020-04-27
- added methods to continously move text on either X or Y axis
- added author sprite in the game menu

## [0.7] - 2020-04-26
- added game title in the menu
- added support for level up in the game, missile count will increase every level
- added sound for level up
- removed redundant condition check in jet sprite
- renamed GameInputText to GameMenuText

## [0.6] - 2020-04-26
- refactored game code into modules based on sprite
- introduced game environment to maintain and share game environment data across different various modules

## [0.5] - 2020-04-26
- added support for continous jet movement towards cursor; even on fixed mouse position

## [0.4] - 2020-04-26
- added suport for mouse movement as game input
- added input mode screen on game start
- fixed replay text sprite rendering

## [0.3] - 2020-04-25
- added support for game replay

## [0.2] - 2020-04-24
- refactored code and added comments
- changed jet image and collision sound
- changed cloud image and added support for random cloud design

## [0.1] - 2020-04-23
- first version with default game play
- added gameover message during collision
- added game scoring and playtime


bluesky.py
# Import the pygame module
import pygame
import random
from game.data import InputMode
from game.environment import GameEnvironment
from game.sprites.jet import Jet
from game.sprites.missile import Missile
from game.sprites.cloud import Cloud
from game.sprites.star import Star
from game.sprites.text import Text
from game.sprites.text.score import ScoreText
from game.sprites.text.replay import ReplayText
from game.sprites.text.gamemenu import GameMenuText

__name__ = 'Bluesky'            # setting game name    
version = '0.12'               # setting game version

def play_bluesky():
    pygame.mixer.init()                                                 # initializing same audio mixer with default settings
    pygame.init()                                                       # initializing pygame
    game_env = GameEnvironment()                                        # initializing game environment

    game_env.variables.moveup_sound.set_volume(0.1)
    game_env.variables.movedown_sound.set_volume(0.1)
    game_env.variables.collision_sound.set_volume(1.5)
    game_env.variables.levelup_sound.set_volume(1.5)
    game_env.variables.shoot_sound.set_volume(1.5)
    game_env.variables.hit_sound.set_volume(3)
    game_env.variables.powerup_sound.set_volume(10)

    pygame.mixer.music.load(game_env.constants.game_sound.get('music'))     # setting main game background music
    pygame.mixer.music.play(loops=-1)                                       # lopping the main game music
    pygame.mixer.music.set_volume(0.5)

    screen = pygame.display.set_mode((800,600))     # creating game screen with custom width and height
    
    screen = pygame.display.set_mode((game_env.constants.screen_width, game_env.constants.screen_height), game_env.FULLSCREEN)     # creating game screen with custom width and height
    pygame.display.set_caption('{} ver. {}'.format(__name__, version))              # setting name of game window
    pygame.mouse.set_visible(False)                                                     # hiding the mouse pointer from the game screen

    gameclock = pygame.time.Clock()                                                     # setting up game clock to maintain constant fps

    ADD_MISSILE = pygame.USEREVENT + 1                                                  # creating custom event to automatically add missiles in the screen
    pygame.time.set_timer(ADD_MISSILE, int(1000/game_env.constants.missile_per_sec))    # setting event to auto-trigger every 500ms; 2 missiles will be created every second

    ADD_CLOUD = pygame.USEREVENT + 2                                                    # creating custom event to automatically add cloud in the screen
    pygame.time.set_timer(ADD_CLOUD, int(1000/game_env.constants.cloud_per_sec))        # setting event to auto-trigger every 1s; 1 cloud will be created every second

    running = True                                                                      # game running variable
    gameover = False                                                                    # no gameover by default
    game_started = False                                                                # game is not started by default
    game_level = 1                                                                      # game level is set to 1
    game_score = 0                                                                      # default game score
    game_playtime = 0                                                                   # default game playtime
    star_shown = False
    mouse_pos = (game_env.constants.screen_width, game_env.constants.screen_height/2)   # default mouse position, let the jet move forward on a straight line
    screen_color = game_env.constants.background_default if game_started else game_env.constants.background_special

    stars = pygame.sprite.GroupSingle()                                                 # group of stars with max 1 sprite
    clouds = pygame.sprite.Group()                                                      # creating cloud group for storing all the clouds in the game
    missiles = pygame.sprite.Group()                                                    # creating missile group for storing all teh missiles in the game
    deactivated_missile = pygame.sprite.Group()                                                # creating missile group for storing all teh missiles in the game
    welcome_screen_sprites = pygame.sprite.Group()

    gamemenu_sprite = GameMenuText(game_env)                                            # creating GameMenuText sprite
    gametitle_sprite = Text(game_env, "{} {}".format(__name__, version), 100, x_pos=game_env.constants.screen_width/2 , y_pos=100)                                      # creating gametitle_sprite text sprite with game name
    gameauthor_sprite = Text(game_env, "Written by: Lakhya Jyoti Nath (ljnath)", 24, x_pos=game_env.constants.screen_width/2 , y_pos= game_env.constants.screen_height-15)  # creating game author
    gamehelp_sprite = Text(game_env, "(Instructions: Move -> mouse/arrow keys, Shoot -> spacebar/mouseclick)", 22, x_pos=game_env.constants.screen_width/2 , y_pos= 140)  # creating game help
    game_env.variables.noammo_sprite = Text(game_env, "NO AMMO !", 24)                 # creating noammo-sprite 

    welcome_screen_sprites.add(gamemenu_sprite)                                                         # adding GameMenuText to welcome_screen_sprites group
    welcome_screen_sprites.add(gametitle_sprite)                                
    welcome_screen_sprites.add(gameauthor_sprite)
    welcome_screen_sprites.add(gamehelp_sprite)

    [game_env.variables.all_sprites.add(sprite) for sprite in welcome_screen_sprites]                   # adding all welcome_screen_sprites sprite to all_sprites

    jet = Jet(game_env)                                                                 # creating jet sprite
    scoretext_sprite = ScoreText(game_env)                                              # creating scoreboard sprite
    replaytext_sprite = ReplayText(game_env)                                            # creating ReplayText sprite
    gameover_sprite = Text(game_env, "GAME OVER", 60, x_pos=game_env.constants.screen_width/2 - 10 ,y_pos = game_env.constants.screen_height/2-25)                          # creating game over message sprite

    # Main game loop
    while running:
        for event in pygame.event.get():                                                                                        # Look at every event in the queue
            if event.type == game_env.KEYDOWN and event.key == game_env.K_ESCAPE or event.type == game_env.QUIT:                # stopping game when ESC key is pressed or when the game window is closed
                running = False
            elif game_started and not gameover:
                if game_env.variables.game_input ==  InputMode.MOUSE and event.type == game_env.MOUSEMOTION:     # moving jet based on mouse movement
                    mouse_pos = pygame.mouse.get_pos()                                                                              # saving the mouse co-ordinate for smooth movement later
                elif (game_env.variables.game_input ==  InputMode.KEYBOARD and event.type == game_env.KEYDOWN and event.key == game_env.K_SPACE) or (game_env.variables.game_input ==  InputMode.MOUSE and event.type==game_env.MOUSEBUTTONDOWN):
                    jet.shoot()
                if event.type == ADD_MISSILE:                                       # is event to add missile is triggered; missles are not added during gameover
                    new_missile = Missile(game_env)                                 # create a new missile
                    missiles.add(new_missile)                                       # adding the missile to missle group
                    game_env.variables.all_sprites.add(new_missile)                 # adding the missile to all_sprites group as well
            elif not game_started and event.type == game_env.KEYDOWN and event.key == game_env.K_RETURN:                        # checking game input mode
                pygame.mouse.set_visible(True if game_env.variables.game_input == InputMode.MOUSE else False)                   # displaying mouse cursor based on user input mode
                screen_color = game_env.constants.background_default                                                            # restoring screen colot
                game_started = True                                                                                             # starting game               
                [sprite.kill() for sprite in welcome_screen_sprites]                                                                            # kill all the welcome_screen_sprites sprite sprite
                game_env.variables.all_sprites.add(jet)                                                                         # adding the jet to all_sprites
                game_env.variables.all_sprites.add(scoretext_sprite)                                                            # adding the scoreboard to all_sprites
            if gameover and event.type == game_env.KEYDOWN and event.key == game_env.K_RETURN and game_started: # checking for replay text only after the game is started
                if replaytext_sprite.replay_choice:
                    gameover = False                                            # setting gameover variable to false as user as opted to replay
                    jet = Jet(game_env)                                         # re-creating the jet
                    game_env.variables.ammo = 100
                    game_env.variables.all_sprites = pygame.sprite.Group()      # re-creating group of sprites 
                    game_env.variables.all_sprites.add(jet)                     # adding the jet to all_sprites
                    game_env.variables.all_sprites.add(scoretext_sprite)        # adding the scoreboard to all_sprites
                    clouds.empty()                                              # empting the cloud group
                    missiles.empty()                                            # empting the missle group
                    screen_color = game_env.constants.background_default        # restoring  screen color
                    game_score, game_playtime, game_level = 0, 0, 1             # reseting game score, playtime and game_level
                    star_shown = False
                    pygame.time.set_timer(ADD_MISSILE, int(1000/game_env.constants.missile_per_sec))
                else:
                    running = False                                             # stopping game as user as opted not to replay
            if event.type == ADD_CLOUD:
                new_cloud = Cloud(game_env)                                     # is event to add cloud is triggered
                clouds.add(new_cloud)                                           # create a new cloud
                game_env.variables.all_sprites.add(new_cloud)                   # adding the cloud to all_sprites group
                if not gameover and game_started:
                    game_playtime += 1                                          # increasing playtime by 1s as this event is triggered every second; just reusing existing event instead of recreating a new event
                    if not star_shown and random.randint(0,30) % 3 == 0:        # probabity of getting a star is 30%
                        star = Star(game_env)
                        stars.add(star)
                        game_env.variables.all_sprites.add(star)
                        star_shown = True
                    if game_playtime % 20 == 0:                                 # changing game level very 20s
                        star_shown = False
                        game_env.variables.levelup_sound.play()                 # playing level up sound
                        game_level += 1                                         # increasing the game level
                        pygame.time.set_timer(ADD_MISSILE, int(1000/(game_env.constants.missile_per_sec + game_level))) # updating timer of ADD_MISSLE for more missiles to be added
                        game_env.variables.ammo += 50                                                                   # adding 50 ammo on each level up
                        game_env.variables.all_sprites.remove(game_env.variables.noammo_sprite)                         # removing no ammo sprite when ammo is refilled
                        

        screen.fill(screen_color)                                                                   # Filling screen with sky blue color
        [screen.blit(sprite.surf, sprite.rect) for sprite in game_env.variables.all_sprites]        # drawing all sprites in the screen

        # missile hit
        if pygame.sprite.spritecollideany(jet, missiles):                                           # Check if any missiles have collided with the player; if so
            jet.kill()                                                                              # killing the jet
            game_env.variables.moveup_sound.stop()
            game_env.variables.movedown_sound.stop()
            game_env.variables.collision_sound.play()
            game_env.variables.all_sprites.add(gameover_sprite)                                     # adding gameover text sprite to all_sprites for repetated rendereing incase of gameover
            game_env.variables.all_sprites.add(replaytext_sprite)                                   # adding replay text sprite to all_sprites for repetated rendereing incase of gameover
            game_env.variables.all_sprites.remove(game_env.variables.noammo_sprite)
            gameover = True                                                                         # setting gameover to true to prevent new missiles from spawning
        
        # missile hit
        collision = pygame.sprite.groupcollide(missiles, game_env.variables.bullets, True, True)    # checking for collision between bullets and missiles, killing each one of them on collision
        if len(collision) > 0:
            game_env.variables.hit_sound.play()                                                     # play missile destroyed sound
            game_score += len(collision) * 10                                                       # 1 missle destroyed = 10 pts.

        # powerup hit
        if pygame.sprite.spritecollideany(jet, stars):                                              # collition between jet and star (powerup)
            game_env.variables.powerup_sound.play()
            [game_env.variables.all_sprites.remove(s) for s in stars.sprites()]                     # removing the star from all_sprites to hide from screen
            game_score += 100 * game_level                                                                      # increasing game score by 100
            stars.empty()                                                                           # removing star from stars group
            for missile in missiles.sprites():                                                      
                missile.deactivate()                                                                # making missile as deactivated
                deactivated_missile.add(missile)                                                    # adding missile to deactivated_missile group
                missiles.remove(missile)                                                            # remove missiles from missles group to avoid collision with jet 
            
        pygame.display.flip()                                                                       # updating display to the screen
        gameclock.tick(game_env.constants.fps)                                                      # ticking game clock at 30 to maintain 30fps

        pressed_keys = pygame.key.get_pressed()                                                     # getting all the pressed keys
        if game_started and not gameover and game_env.variables.game_input == InputMode.KEYBOARD:
            jet.update(pressed_keys)                                                                # calling update() to act according to pressed keys
        elif game_started and gameover:
            replaytext_sprite.update(pressed_keys)                                                  # allowing user to select game replay option during gameover mode
        elif not game_started:
            gamemenu_sprite.update(pressed_keys)                                                    # allowing user to select game input type when the game is not started
            gameauthor_sprite.moveOnXaxis(2)                                                        # moving the game author sprite across the X axis   
        elif game_started and not gameover and game_env.variables.game_input == InputMode.MOUSE:    # performing the jet movement here for smooth movement till mouse cursor
            jet.auto_move(mouse_pos)

        game_env.variables.bullets.update()
        missiles.update()                                                       # update the position of the missiles
        deactivated_missile.update()
        clouds.update()                                                         # update the postition of the clouds
        stars.update()
        scoretext_sprite.update(game_level, game_playtime, game_score)          # update the game score

    pygame.mixer.music.stop()                                                   # stopping game music
    pygame.mixer.quit()                                                         # stopping game sound mixer


if __name__ == 'Bluesky':  
    play_bluesky()                                                              # starting game